### PR TITLE
RC receiver and failsafe cleanup - stage 2

### PR DESCRIPTION
**Pull request by digitalentity**
Better decoupling between failsafe and receiver code. Receiver code made independent from RC provider. Signal loss detection is now done by provider, signal timeout detection - by RC code. FAILSAFE flight mode bypasses the timeout and triggers failsafe instantly now.

### CODE REVIEW

**Comment by digitalentity** [comment]
Bench tested with S.Bus receiver - seems to work fine. @fiam @DzikuVx @martinbudden I'd like to have your oppinion on this.

File: src/main/fc/fc_msp.c [hunk 1]
```diff
@@ -944,6 +944,7 @@ static bool mspFcProcessOutCommand(uint16_t cmdMSP, sbuf_t *dst, mspPostProcessF
         sbufWriteU8(dst, rxConfig()->rx_spi_protocol);
         sbufWriteU32(dst, rxConfig()->rx_spi_id);
         sbufWriteU8(dst, rxConfig()->rx_spi_rf_channel_count);
+        sbufWriteU8(dst, rxConfig()->receiverType);
```
Comment (fiam) [inline src/main/fc/fc_msp.c]: Won't this cause problems if the receiver of the MSP response is expecting a specific length?

File: src/main/fc/settings.yaml [hunk 2]
```diff
@@ -18,6 +18,9 @@ tables:
   - name: pitot_hardware
     values: ["NONE", "AUTO", "MS4525", "ADC", "VIRTUAL", "FAKE"]
     enum: pitotSensor_e
+  - name: receiver_type
+    values: ["NONE", "PWM", "PPM", "SERIAL", "MSP", "SPI"]
+    enum: rxReceiverType_e
```
Comment (fiam) [inline src/main/fc/settings.yaml:23]: Not that there's a problem with this. Just reminds me that I should get around to implementing enum validation in the settings generator.

File: src/main/flight/failsafe.h [hunk 3]
```diff
@@ -129,9 +129,10 @@ typedef enum {
 
 typedef struct failsafeState_s {
     int16_t events;
-    bool monitoring;
-    bool active;
-    bool controlling;
+    bool monitoring;                        // Flag that failsafe is monitoring RC link
+    bool suspended;                         // Failsafe is temporary suspended
+    bool active;                            // Failsafe is active (on RC link loss)
+    bool controlling;                       // Failsafe is driving the sticks instead of pilot
```
Comment (fiam) [inline src/main/flight/failsafe.h]: I appreciate the comments here. We're really missing comments in a bunch of places that would help potential new developers a lot. Could you also add under which conditions `suspended` could be true?

File: src/main/rx/rx.h [hunk 4]
```diff
@@ -90,11 +103,12 @@ typedef struct rxChannelRangeConfig_s {
 PG_DECLARE_ARRAY(rxChannelRangeConfig_t, NON_AUX_CHANNEL_COUNT, rxChannelRangeConfigs);
 
 typedef struct rxConfig_s {
+    uint8_t receiverType;
     uint8_t rcmap[MAX_MAPPABLE_RX_INPUTS];  // mapping of radio channels to internal RPYTA+ order
-    uint8_t serialrx_provider;              // type of UART-based receiver (0 = spek 10, 1 = spek 11, 2 = sbus). Must be enabled by FEATURE_RX_SERIAL first.
+    uint8_t serialrx_provider;              // type of UART-based receiver (0 = spek 10, 1 = spek 11, 2 = sbus). Only used if recevierType is RX_TYPE_SERIAL
```
Comment (fiam) [inline src/main/rx/rx.h]: Typo in comment `recevierType` -> `receiverType`. Also, comment seems outdated. Better to change the type of `serialrx_provider` to `SerialRXType` and that way the first sentence of the comment can go away.

File: src/main/rx/rx.h [hunk 5]
```diff
@@ -90,11 +103,12 @@ typedef struct rxChannelRangeConfig_s {
...
     uint8_t sbus_inversion;                 // default sbus (Futaba, FrSKY) is inverted. Support for uninverted OpenLRS (and modified FrSKY) receivers.
     uint8_t halfDuplex;                     // allow rx to operate in half duplex mode on F4, ignored for F1 and F3.
-    uint8_t rx_spi_protocol;               // type of nrf24 protocol (0 = v202 250kbps). Must be enabled by FEATURE_RX_NRF24 first.
+    uint8_t rx_spi_protocol;                // type of SPI receiver protocol. Only used if receiverType is RX_TYPE_SPI
```
Comment (fiam) [inline src/main/rx/rx.h]: Could we use an enum for `rx_spi_protocol`?

File: src/main/target/COLIBRI/config.c [hunk 6]
```diff
@@ -33,6 +33,6 @@ void targetConfiguration(void)
 {
     mixerConfigMutable()->mixerMode = MIXER_HEX6X;
     rxConfigMutable()->serialrx_provider = 2;
```
Comment (fiam) [inline src/main/target/COLIBRI/config.c]: Would be better to use the constant in the `SerialRxType` enum rather than a magic `2`.

**Review by fiam — COMMENTED** [review]
Just added a few minor comments. Overall I really like the idea of removing the RX type from the features, it never felt like it belonged there. The failsafe cleanup is also nice, makes the code easier to read.

File: src/main/fc/fc_msp.c (hunk 1 above)
Comment (digitalentity) [reply src/main/fc/fc_msp.c]: Not if other party (configurator) checks for firmware version.

File: src/main/rx/rx.h (hunk 5 above)
Comment (digitalentity) [reply src/main/rx/rx.h]: `sizeof` of enum is not defined, while we expect this to occupy exactly 1 byte

Comment (fiam) [reply src/main/rx/rx.h]: Yup, I had not thought of that. Can you name the enum with the valid values in the comment then?

File: src/main/common/maths.c [hunk 7]
```diff
@@ -16,6 +16,7 @@
  */
```
Comment (fiam) [inline src/main/common/maths.c:16]: Looks like you included the changes in maths.{c,h} by mistake

Comment (digitalentity) [reply src/main/common/maths.c:16]: No mistake here. I need a median filter for int16_t

**Comment by digitalentity** [comment]
Tested for I-Bus, S.Bus and PPM receivers. No issues.

**Comment by stronnag** [comment]
Bench tested for Specktrum. All inputs appear correct in the configurator, and failsafe is indicated when the TX is powered-off.

**Comment by digitalentity** [comment]
~~Merging~~
We need a Configurator update to handle new receiver types...

**Comment by stronnag** [comment]
Flight tested on multiple hardware (Matek F405, SPF3EVO, Omnibus F3, Dodo(SPF3)). No problems, at least until the dev merge resulting in https://github.com/iNavFlight/inav/issues/2154 -- which is not a problem from this branch.

**Comment by fiam** [comment]
@digitalentity I'd apply this patch on top of this one. ßF has added another parameter after `rx_spi_rf_channel_count` for `fpvCamAngleDegrees`, so I'd insert a zero between `rx_spi_rf_channel_count` and `receiverType`. Eventually we might want to support `fpvCamAngleDegrees`. Then we can fix the issue in the configurator ignoring `rx_spi_rf_channel_count` and we're good to go.

```diff
diff --git a/src/main/fc/fc_msp.c b/src/main/fc/fc_msp.c
index f145c79c..22b227fb 100644
--- a/src/main/fc/fc_msp.c
+++ b/src/main/fc/fc_msp.c
@@ -683,13 +683,15 @@ static bool mspFcProcessOutCommand(uint16_t cmdMSP, sbuf_t *dst, mspPostProcessF
         sbufWriteU8(dst, rxConfig()->spektrum_sat_bind);
         sbufWriteU16(dst, rxConfig()->rx_min_usec);
         sbufWriteU16(dst, rxConfig()->rx_max_usec);
-        sbufWriteU8(dst, 0); // for compatibility with betaflight
-        sbufWriteU8(dst, 0); // for compatibility with betaflight
-        sbufWriteU16(dst, 0); // for compatibility with betaflight
+        sbufWriteU8(dst, 0); // for compatibility with betaflight (rcInterpolation)
+        sbufWriteU8(dst, 0); // for compatibility with betaflight (rcInterpolationInterval)
+        sbufWriteU16(dst, 0); // for compatibility with betaflight (airModeActivateThreshold)
         sbufWriteU8(dst, rxConfig()->rx_spi_protocol);
         sbufWriteU32(dst, rxConfig()->rx_spi_id);
         sbufWriteU8(dst, rxConfig()->rx_spi_rf_channel_count);
+        sbufWriteU8(dst, 0); // for compatibility with betaflight (fpvCamAngleDegrees)
         sbufWriteU8(dst, rxConfig()->receiverType);
         break;
 
     case MSP_FAILSAFE_CONFIG:
@@ -1740,13 +1742,15 @@ static mspResult_e mspFcProcessInCommand(uint16_t cmdMSP, sbuf_t *src)
         sbufReadU8Safe(&rxConfigMutable()->spektrum_sat_bind, src);
         sbufReadU16Safe(&rxConfigMutable()->rx_min_usec, src);
         sbufReadU16Safe(&rxConfigMutable()->rx_max_usec, src);
-        sbufReadU8Safe(NULL, src); // for compatibility with betaflight
-        sbufReadU8Safe(NULL, src); // for compatibility with betaflight
-        sbufReadU16Safe(NULL, src); // for compatibility with betaflight
+        sbufReadU8Safe(NULL, src); // for compatibility with betaflight (rcInterpolation)
+        sbufReadU8Safe(NULL, src); // for compatibility with betaflight (rcInterpolationInterval)
+        sbufReadU16Safe(NULL, src); // for compatibility with betaflight (airModeActivateThreshold)
         sbufReadU8Safe(&rxConfigMutable()->rx_spi_protocol, src);
         sbufReadU32Safe(&rxConfigMutable()->rx_spi_id, src);
         sbufReadU8Safe(&rxConfigMutable()->rx_spi_rf_channel_count, src);
+        sbufReadU8Safe(NULL, src); // for compatibility with betaflight (fpvCamAngleDegrees)
         sbufReadU8Safe(&rxConfigMutable()->receiverType, src);              // Won't be modified if buffer is not large enough
         break;
 
     case MSP_SET_FAILSAFE_CONFIG:
diff --git a/src/main/rx/rx.c b/src/main/rx/rx.c
index b48ee8bf..2ea1e6d6 100755
--- a/src/main/rx/rx.c
+++ b/src/main/rx/rx.c
@@ -373,7 +373,7 @@ void calculateRxChannelsAndUpdateFailsafe(timeUs_t currentTimeUs)
     rxLastUpdateTimeUs = currentTimeUs + DELAY_50_HZ;
 
     rxFlightChannelsValid = true;
-    
+
     // Read and process channel data
     for (int channel = 0; channel < rxRuntimeConfig.channelCount; channel++) {
         const uint8_t rawChannel = calculateChannelRemapping(rxConfig()->rcmap, REMAPPABLE_CHANNEL_COUNT, channel);
```

**Comment by fiam** [comment]
Configurator support is done for this one. PR at https://github.com/iNavFlight/inav-configurator/pull/267

**Comment by stronnag** [comment]
I flashed this with the intent to fly it today. Then I saw that is has reversed status bits, as if #2154 had never happened. So I re-flashed in the field [0] (how sad) and flew development instead, which was as usual, stable and awesome.

Back home, I read `git log` and see the last merge with development was an age ago. Maybe tomorrow I am less of a wimp, especially as current dev merges cleanly.
 
[0] Whatever, iNav's magnificent feature of being able to save / restore acc calibration makes this possible, as there is not otherwise a flat surface in sight.

**Comment by digitalentity** [comment]
@stronnag yes, last catch-up with development was a while ago. Just fixed that. Everything seems to be working properly.

**Comment by digitalentity** [comment]
I'm merging this now. Configurator PR is ready to be merged as well - thanks to @fiam :+1:

**Comment by stronnag** [comment]
I flew this today without any problems.
Don't forget to mention in the release note; regardless we shall have a load of 1.8 issues "tx does not connect, worked in 1.73". Maybe we need a "Release Note" tag for PRs to help us remember the significant items.

**Comment by fiam** [comment]
There's quite a big list of things to mention. We should start going to the 1.8 milestone next week and create a wiki page to list all changes. There are also a bunch of new and renamed settings, so dumps won't restore cleanly.

**Pull request merged**